src/contexts/ApiContext.tsx

```
@@ -0,0 +1,68 @@
+import { createContext, PropsWithChildren, useReducer } from "react";
+
+type State = {
+  data: Record<string, unknown>;
+  loadingStates: Record<string, boolean>;
+  errorStates: Record<string, unknown>;
+};
+
+type Action =
+  | { type: "startFetch"; name: string }
+  | { type: "fetchSuccess"; name: string; payload: unknown }
+  | { type: "fetchError"; name: string; error: unknown }
+  | { type: "endFetch"; name: string };
+
+const initialState: State = {
+  data: {},
+  loadingStates: {},
+  errorStates: {},
+};
+
+function apiReducer(state: State, action: Action): State {
+  switch (action.type) {
+    case "startFetch":
+      return {
+        ...state,
+        loadingStates: { ...state.loadingStates, [action.name]: true },
+        errorStates: { ...state.errorStates, [action.name]: undefined },
+      };
+    case "fetchSuccess":
+      return {
+        ...state,
+        data: { ...state.data, [action.name]: action.payload },
+        loadingStates: { ...state.loadingStates, [action.name]: false },
+        errorStates: { ...state.errorStates, [action.name]: undefined },
+      };
+    case "fetchError":
+      return {
+        ...state,
+        loadingStates: { ...state.loadingStates, [action.name]: false },
+        errorStates: { ...state.errorStates, [action.name]: action.error },
+      };
+    case "endFetch":
+      return {
+        ...state,
+        loadingStates: { ...state.loadingStates, [action.name]: false },
+      };
+    default:
+      return state;
+  }
+}
+
+export const ApiContext = createContext<{
+  state: State;
+  dispatch: React.Dispatch<Action>;
+}>({
+  state: initialState,
+  dispatch: () => {},
```

이렇게 뒀을때, 실수로 context를 provider 없이 사용하면 런타임 오류 발생할거 같아요

---

어떤 의도로 피드백을 주신 건지 이해가 잘 가지 않습니다.
혹시 힌트나 자세한 설명을 좀 더 주실 수 있을까요??

```ts
export function useData<T>({
  fetcher,
  name,
}: {
  fetcher: () => Promise<T>;
  name: FetchActionName;
}) {
  const context = useContext(ApiContext);
  if (!context) {
    throw new Error("apiContext는 apiContextProvider 내부에 위치해야 합니다.");
  }
...
```
여기서 throw new Error를 하고 있어서 문제라는 걸까요?? 👀

---

아하 요 코드에서 제가 생각한 부분은,

useContext(ApiContext)를 사용할 때, ApiProvider 없이 사용하면 이 기본값이 그대로 사용됩니다. 이때 dispatch()를 호출해도 아무 일도 안 일어납니다. 심지어 오류도 발생하지 않아요. 즉, silent fail이 됩니다.

코드로 보면 좀더 쉽게 이해가 될텐데,
```
function App() {
  return <CartPage />; // ApiProvider 없이 사용
}
```
이런코드에서 dispatch()가 실행되지만 실제 리듀서가 동작하지 않고, 콘솔 에러도 없습니다.
즉, 버그가 발생해도 디버깅이 매우 어렵습니다.

라는 생각으로 달았었습니다! 
근데 사실 요 로직은 로건이 보여준 코드로 해결됩니다!
```
const context = useContext(ApiContext);
  if (!context) {
    throw new Error("apiContext는 apiContextProvider 내부에 위치해야 합니다.");
  }
```
이렇게 잘 처리해주고 있는데, 이 코드 덕분에 useData에서 context가 undefined일 때 명시적으로 에러를 던지기 때문에, 위에서 말한 silent fail 문제가 발생하지 않습니다.

여기에 대해서 잘 이해하고 있는지 궁금하기도 하고 서로 싱크가 잘 맞혀지는지 확인차 코멘트를 달아봤어용